blog/_posts/2016-12-xx-moredots.md

```diff
@@ -0,0 +1,456 @@
+so many dots, such as you might use in Julia 0.4 or in other languages
+(most famously Matlab, Python/Numpy, or R).   
+
+```jl
+X = f(2 * X.^2 + 6 * X.^3 - sqrt(X))
+```
+
+Of course, this assumes that the functions `sqrt` and `f` are "vectorized,"
+i.e. that they accept vector arguments `X` and compute the
+function elementwise.  This is true of `sqrt` in Julia 0.4, but it
+means that we have to rewrite our function `f` from above in a vectorized style, as
+e.g. `f(x) = 3x.^2 + 5x + 2` (changing `f` to use the elementwise operator `.^` because
+`array^scalar` is not defined).   (If we cared a lot about efficiency, we
+might instead define a special method `f(X::AbstractArray) = map(f, X)`
+or use the `@vectorize_1arg f Number` macro defined in Julia 0.4.)
+
+### Which functions are vectorized?
+
+As an aside, this example illustrates an annoyance with the vectorized style:
+you have to *decide in advance* whether a given function `f(x)`
+will also be applied elementwise to arrays, and either
+write it specially or define a corresponding elementwise method.
+
+(Even if a function *accepts* an array argument `x` — our function `f`
+accepts any `x` type, and in Matlab or R there is no distinction between
+a scalar and a 1-element array — that doesn't mean it will *work* elementwise
+for an array unless you write the function with that in mind.)
+
+For library functions like `sqrt`, this means that the library authors
+have to guess at which functions should have vectorized methods, and users
+have to guess at what vaguely defined subset of library functions work
+for vectors.
+
+One possible solution is to vectorize *every function automatically*.   The
+language [Chapel](https://en.wikipedia.org/wiki/Chapel_%28programming_language%29)
+does this: every function `f(x...)` implicitly
+defines a function `f(x::Array...)` that evaluates `map(f, x...)`
+[(Chamberlain et al, 2011)](http://pgas11.rice.edu/papers/ChamberlainEtAl-Chapel-Iterators-PGAS11.pdf).
+This could be implemented in Julia as well via
+function-call overloading [(Bezanson, 2015: chapter 4)](https://github.com/JeffBezanson/phdthesis/blob/master/main.pdf),
+but we chose to go in a different direction.
+
+Instead, starting in Julia 0.5, *any* function `f(x)` can be applied elementwise
+to an array `X` with the ["dot call" syntax `f.(X)`](http://docs.julialang.org/en/stable/manual/functions/#dot-syntax-for-vectorizing-functions).
+Thus, the *caller* decides which functions to vectorize.  In Julia 0.6,
+"traditionally" vectorized library functions like `sqrt(X)` are [deprecated](https://github.com/JuliaLang/julia/pull/17302) in
+favor of `sqrt.(X)`, and dot operators
+like `x .+ y` are [now equivalent](https://github.com/JuliaLang/julia/pull/17623) to
+dot calls `(+).(x,y)`.   (Unlike Chapel's implicit vectorization, Julia's
+`f.(x...)` syntax corresponds to `broadcast(f, x...)` rather than `map`,
+which allows you to combine arrays of different shapes/dimensions.)
```

Comment:
Whether the "which" applies to `broadcast(f, x...)` or `map` seems ambiguous. Perhaps "which allows" -> "allowing" or "rather than `map`" -> ("rather than `map`")?